File: deep_search-dev/deepsearch/main_search.py
import re
import time
from deepsearch.prompt.prompt_lang import related_url_prompt
from api.llm_api import llm
from api.search_api import SearchAPI
from api.crawler_api import CrawlerAPI


class SearchCrawl():

    def __init__(self, topic, rewrite_query, summary_search, logger):
        """
        Args:
            topic: 主题
            logger: 日志
            rewrite_query: 子主题
            query_type: 子主题对应搜索时间范围
            summary_search: 总结的搜索结果
        """
        self.origin_query = topic
        self.rewrite_query = rewrite_query
        self.logger = logger
        self.summary_search = summary_search
        self.all_searched_urls = []  # 保存所有搜索到的URL
        self.selected_urls = []  # 保存已经挑选过的URL

    async def run(self):
        """
        1、搜索
        2、模型选出最适合的 url
        3、爬取具体内容

        Returns:
            爬取结果的列表
        """
        self.logger.info(f'开始搜索爬取...')

        # 搜索，返回url、标题、摘要、评分
        tmp_start_time = time.time()
        self.logger.info(
            f'开始调用搜索接口...\ntopic: {self.origin_query}\nrewrite_query: {self.rewrite_query}')
        
        # 创建SearchAPI实例并调用搜索接口
        search_api = SearchAPI(logger=self.logger)
        search_result = search_api.search(self.rewrite_query)
        self.logger.info(f'搜索接口调用工作流: {search_result.get('debug_url', '')}')
        
        if search_result.get("msg") == "success":
            serach_res_list = search_result.get("results", [])
        else:
            self.logger.error(f'搜索接口调用失败: {search_result.get("msg", "")}')
            serach_res_list = [] 
        if len(serach_res_list) == 0:
            self.logger.warning(f'共搜索到 0 条结果...')
            return []
        else:
            self.logger.info(f'共搜索到 {len(serach_res_list)} 条结果...')
        self.logger.info(f'调用搜索接口耗时：{time.time() - tmp_start_time}s')

        # 将新搜索到的URL添加到all_searched_urls列表中
        current_search_urls = [search.get("url", "") for search in serach_res_list]
        self.all_searched_urls.extend(current_search_urls)
        self.logger.info(f'当前总搜索URL数量: {len(self.all_searched_urls)}')

        # 根据搜索结果，用大模型进行url挑选
        tmp_start_time = time.time()
        self.logger.info(f'开始进行 url 筛选...')
        to_fetch_urls = await self.select_related_url(serach_res_list, self.logger)
        self.logger.info(f'共挑出 {len(to_fetch_urls)} 个需要的 url: {to_fetch_urls}')
        self.logger.info(f'调用大模型筛选 url 耗时：{time.time() - tmp_start_time}s')

        # 爬取url
        tmp_start_time = time.time()
        self.logger.info(f'开始爬取 url...')
        
        # 使用CrawlerAPI爬取URL
        async with CrawlerAPI(logger=self.logger) as crawler:
            fetch_res = await crawler.crawl_urls(to_fetch_urls)
        
        self.logger.info(f'爬取完成...')
        self.logger.info(f'调用爬虫接口获取网页内容耗时：{time.time() - tmp_start_time}s')

        if not fetch_res:
            self.logger.warning(f'搜索爬取结果为空！')

        # 爬取结果文本合并
        crawl_res = [data.get("content", "").replace('\n', '').replace('\t', '') for data in fetch_res if data.get("content", "") != 'NA'] if fetch_res else []
        return crawl_res

    async def select_related_url(self, serach_res_list, logger):
        """
        选择与 topic 需要的链接

        Args:
            serach_res_list: 搜索结果
            logger: 日志

        Returns:
            选出结果后搜索结果原始列表
        """
        # 组装搜索结果给模型使用
        ori_urls = []  # 兜底
        text = ''
        url_map = {}
        
        # 从all_searched_urls中排除已经挑选过的URL
        available_urls = [url for url in self.all_searched_urls if url not in self.selected_urls]
        
        # 构建URL映射和文本
        for i, search in enumerate(serach_res_list, start=1):
            # 从搜索结果中提取信息
            summary = search.get("summary", "")
            title = search.get("title", "")
            url = search.get("url", "")
            
            # 只处理可用的URL
            if url in available_urls:
                # 将信息添加到映射和原始URL列表
                ori_urls.append(url)
                url_map[f'url_{i}'] = url
                text += f'摘要：{summary}\n标题：{title}\n链接：url_{i}\n'
                text += '==============================================================\n'
        
        logger.info(f'映射后链接：{url_map}')
        logger.info(f'可用链接数量: {len(available_urls)}')
        logger.info(f'已挑选链接数量: {len(self.selected_urls)}')

        # 选择
        prompt = related_url_prompt.format(text=text, topic=self.origin_query, summary_search=self.summary_search).strip()
        logger.info(f'选择相关的搜索结果 prompt:\n{prompt}')
        response = await llm.infer(prompt)
        logger.info(f'大模型选择需要的搜索结果结果：\n{response}')

        url_lst = re.findall(r'```python(.*?)```', str(response), flags=re.S)
        if url_lst:
            related_urls = eval(url_lst[0])
            related_urls = [url_map[url_] for url_ in related_urls if url_ in url_map]
            # 更新已挑选的URL列表
            self.selected_urls.extend(related_urls)
        else:
            logger.warning(f'解析结果失败，使用原始所有结果...')
            related_urls = ori_urls
            # 更新已挑选的URL列表
            self.selected_urls.extend(related_urls)

        return related_urls

File: Agent/deep_search-dev-go/deepsearch/main_search.go
package deepsearch

import (
	"context"
	"deepsearch/api"
	"deepsearch/deepsearch/prompt"
	"deepsearch/logs"
	"encoding/json"
	"fmt"
	"strings"
)

// SearchCrawl 搜索和爬取结构体
type SearchCrawl struct {
	SearchAPI  *api.SearchAPI
	CrawlerAPI *api.CrawlerAPI
	Logger     *logs.Logger
}

// NewSearchCrawl 创建搜索和爬取实例
func NewSearchCrawl(logger *logs.Logger) *SearchCrawl {
	return &SearchCrawl{
		SearchAPI:  api.NewSearchAPI(logger),
		CrawlerAPI: api.NewCrawlerAPI(logger),
		Logger:     logger,
	}
}

// Run 运行搜索和爬取：返回 crawlResList
func (sc *SearchCrawl) Run(ctx context.Context, topic string, haveQuery []string, allSummarySearch []string) ([]string, error) {
	// 初始化爬虫API
	err := sc.CrawlerAPI.InitBrowser()
	if err != nil {
		sc.Logger.Errorf("初始化爬虫API失败: %v", err)
		return nil, fmt.Errorf("初始化爬虫API失败: %v", err)
	}
	defer sc.CrawlerAPI.CloseBrowser()

	// 执行搜索
	sc.Logger.Infof("开始调用搜索接口...\ntopic: %s\nrewrite_query: %v", topic, haveQuery)
	searchResults, err := sc.SearchAPI.Search(haveQuery)
	if err != nil {
		sc.Logger.Errorf("搜索失败: %v", err)
		return nil, fmt.Errorf("搜索失败: %v", err)
	}

	sc.Logger.Infof("搜索接口调用工作流: %s", searchResults.DebugURL)
	
	if len(searchResults.Results) == 0 {
		sc.Logger.Warning("共搜索到 0 条结果...")
		return []string{}, nil
	}
	sc.Logger.Infof("共搜索到 %d 条结果...", len(searchResults.Results))

	// 选择相关URL
	sc.Logger.Info("开始进行 url 筛选...")
	relatedURLs, err := sc.selectRelatedURLs(ctx, topic, searchResults, allSummarySearch)
	if err != nil {
		sc.Logger.Errorf("选择相关URL失败: %v", err)
		return nil, fmt.Errorf("选择相关URL失败: %v", err)
	}

	sc.Logger.Infof("共挑出 %d 个需要的 url: %v", len(relatedURLs), relatedURLs)

	// 爬取URL内容
	sc.Logger.Info("开始爬取 url...")
	crawlResults, err := sc.CrawlerAPI.CrawlURLs(ctx, relatedURLs)
	if err != nil {
		sc.Logger.Errorf("爬取URL内容失败: %v", err)
		return nil, fmt.Errorf("爬取URL内容失败: %v", err)
	}

	sc.Logger.Info("爬取完成...")

	// 提取爬取内容
	var contents []string
	for _, result := range crawlResults {
		// 清理内容，移除换行符和制表符
		content := strings.ReplaceAll(result.Content, "\n", "")
		content = strings.ReplaceAll(content, "\t", "")
		if content != "" && content != "NA" {
			contents = append(contents, content)
		}
	}

	if len(contents) == 0 {
		sc.Logger.Warning("搜索爬取结果为空！")
	}

	return contents, nil
}

// selectRelatedURLs 选择相关URL
func (sc *SearchCrawl) selectRelatedURLs(ctx context.Context, topic string, searchResults *api.SearchResponse, allSummarySearch []string) ([]string, error) {
	// 构建URL映射
	urlMap := make(map[string]string)
	var searchResultsText []string
	var originalURLs []string // 兜底
	
	for i, result := range searchResults.Results {
		urlKey := fmt.Sprintf("url_%d", i+1) // 从1开始计数，与Python版本一致
		urlMap[urlKey] = result.URL
		originalURLs = append(originalURLs, result.URL)
		
		// 格式化搜索结果文本，与Python版本一致
		resultText := fmt.Sprintf("摘要：%s\n标题：%s\n链接：%s", result.Summary, result.Title, urlKey)
		searchResultsText = append(searchResultsText, resultText)
		searchResultsText = append(searchResultsText, "==============================================================")
	}
	
	sc.Logger.Infof("映射后链接：%v", urlMap)
	
	// 构建提示词，使用Python版本的格式
	promptText := fmt.Sprintf(prompt.RelatedUrlPrompt, topic, allSummarySearch, strings.Join(searchResultsText, "\n"))
	
	sc.Logger.Infof("选择相关的搜索结果 prompt:\n%s", promptText)
	
	// 调用大模型
	llm := api.GlobalLLM
	if llm == nil {
		sc.Logger.Error("全局LLM实例未初始化")
		return nil, fmt.Errorf("全局LLM实例未初始化")
	}
	
	response, err := llm.Infer(ctx, promptText, false, 0.7)
	if err != nil {
		sc.Logger.Errorf("调用大模型失败: %v", err)
		return nil, fmt.Errorf("调用大模型失败: %v", err)
	}
	
	sc.Logger.Infof("大模型选择需要的搜索结果结果：\n%s", response)
	
	// 解析响应，提取URL
	var selectedURLs []string
	response = strings.TrimSpace(response)
	
	// 直接尝试解析JSON数组
	var urlKeys []string
	if err := json.Unmarshal([]byte(response), &urlKeys); err == nil {
		// 成功解析JSON数组
		for _, urlKey := range urlKeys {
			if url, exists := urlMap[urlKey]; exists {
				selectedURLs = append(selectedURLs, url)
			}
		}
	} else {
		// 如果不是有效的JSON，尝试手动解析
		sc.Logger.Warningf("无法解析JSON响应，尝试手动解析: %s", response)
		// 简单的字符串解析，提取url_x格式的字符串
		lines := strings.Split(response, "\n")
		for _, line := range lines {
			line = strings.TrimSpace(line)
			// 移除可能的引号和逗号
			line = strings.Trim(line, "\"'`,")
			if strings.HasPrefix(line, "url_") {
				if url, exists := urlMap[line]; exists {
					selectedURLs = append(selectedURLs, url)
				}
			}
		}
		
		// 如果仍然没有找到URL，使用原始所有结果作为兜底
		if len(selectedURLs) == 0 {
			sc.Logger.Warning("解析结果失败，使用原始所有结果...")
			selectedURLs = originalURLs
		}
	}
	
	if len(selectedURLs) == 0 {
		sc.Logger.Warning("未选择到任何相关URL")
	}
	
	return selectedURLs, nil
}
